### PR TITLE
create backups via UI

### PR DESCRIPTION
**Pull request by machisuji**
WP [#35645](https://community.openproject.org/projects/saas/work_packages/35645/activity)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -37,6 +37,10 @@ gem 'activerecord-session_store', '~> 2.0.0'
 gem 'rails', '~> 6.1.3'
 gem 'responders', '~> 3.0'
 
+# Keep mimemagic at older version until rails provides its own
+# solution without adding a system dependency. We are GPL ourselves so this is fine.
+gem 'mimemagic', git: 'https://github.com/opf/mimemagic', ref: 'bf8d7c2'
```
Comment (oliverguenther) [inline Gemfile]: Can be removed in dev now

File: app/models/export.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Export < ApplicationRecord
+  has_one(
```
Comment (oliverguenther) [inline app/models/export.rb:2]: :+1: for extracting this

File: app/services/concerns/contracted.rb [hunk 3]
```diff
@@ -36,8 +36,8 @@ module Contracted
     attr_accessor :contract_options
 
     def contract_class=(cls)
-      unless cls <= ::BaseContract
-        raise ArgumentError "#{cls.name} is not an instance of BaseContract."
+      unless cls <= ::ModelContract
```
Comment (oliverguenther) [inline app/services/concerns/contracted.rb]: Why did you change this back? I'm pretty sure I originally moved this from ModelContract to BaseContract in case we have a service that doesn't depend on an AR model

File: app/views/admin/backups/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+See docs/COPYRIGHT.rdoc for more details.
+
+++#%>
+
+<% html_title t(:label_administration), t(:backup) -%>
+
+<%= toolbar title: t(:label_backup) %>
+
+<backup
```
Comment (oliverguenther) [inline app/views/admin/backups/show.html.erb]: Just a preference but I'd prefer a

```
tag :backup, data: { 'job-status-id': .... }
```

as that handles escaping the attributes correctly in case you ever want to pass a string with a quote in it

File: frontend/src/app/components/api/op-backup/op-backup.service.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+//++
+
+import {Injectable} from "@angular/core";
+import {HttpClient, HttpEvent, HttpEventType, HttpResponse} from "@angular/common/http";
+import {HalResource} from "core-app/modules/hal/resources/hal-resource";
+import {Observable} from "rxjs";
+import {filter, map, share} from "rxjs/operators";
+import {HalResourceService} from "core-app/modules/hal/services/hal-resource.service";
+
+@Injectable()
```
Comment (oliverguenther) [inline frontend/src/app/components/api/op-backup/op-backup.service.ts]:
```suggestion
@Injectable({ providedIn: 'root' })
```

Instead of providing in the uppermost module angular4 modules, please use this instead. This will result in the service becoming a global singleton whenever any component or service requires it.

File: lib/api/v3/backups/backups_api.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            )
+          end
+          post do
+            current_backup = Backup.last
+
+            if pending_statuses.include? current_backup&.job_status&.status
+              fail ::API::Errors::Conflict, message: "There is already a backup pending."
+            end
+
+            limit = OpenProject::Configuration.backup_daily_limit
```
Comment (oliverguenther) [inline lib/api/v3/backups/backups_api.rb]: I'd probably want to have this moved into the contract level. This ensures it's being validated everywhere and you can react to it with a special response in case of this error symbol.

File: spec/features/admin/backup_spec.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    begin
+      perform_enqueued_jobs
+    rescue StandardError
+      # nothing
+    end
+
+    expect(page).to have_text "The export has completed successfully"
+    expect(subject).to end_with ".zip"
+  end
```
Comment (oliverguenther) [inline spec/features/admin/backup_spec.rb:64]: Missing a spec to test the 429 when the daily limit is being hit

**Review by oliverguenther — CHANGES_REQUESTED** [review]
I got some minor comments.

Additionally, these followings things I noticed:

- [x] The export does not include the attachments even though I requested them through the checkbox
- [x] The export can be triggered by anyone with the permission and doesn't contain the discussed "first-time generated token" thing?

File: app/services/concerns/contracted.rb (hunk 3 above)
Comment (machisuji) [reply app/services/concerns/contracted.rb]: All I did was add a missing comma there after the `ArgumentError` as far as I'm concerned.

Comment (machisuji) [reply app/services/concerns/contracted.rb]: But yeah apparently the class changed too there. Since I based this on release/11.2 originally this might've been mixed up. I'll change the class back.

File: spec/features/admin/backup_spec.rb (hunk 7 above)
Comment (machisuji) [reply spec/features/admin/backup_spec.rb:64]: I've added a spec testing a general (400) error to the feature specs. The specific errros (429, 403 etc.) are covered in the backup API spec.

**Comment by machisuji** [comment]
@oliverguenther Could you please have another look? This now does include the backup token. Though I've changed it so it's not just generated once initially but can be regenerated. Open for discussion I guess but I think the way it is now will give us less trouble (work) later on because people will be able to reset the tokens themselves.

To avoid the idle desktop scenario we talked about a newly generated token cannot be used straight away and also will trigger an email informing the actual user that a new token has been generated. If it wasn't them this gives them the chance to react and reset it again.

**Comment by oliverguenther** [comment]
Functionality-wise, LGTM now. As a user of the feature I'm still a bit confused by the delay as I do not get informed about that when first creating the backup token. Only when I start using it will I be greeted with an error that I still have to activate it.

I'd appreciate it if we extend the info on the page a bit, and also add a piece of documentation about how to trigger the allow_now.

I had a local installation with more than 1GB of attachments, which is why the attachments were not added. This is also not clear to the user and the attachments are just missing. I'd also appreciate either disabling the checkbox or some other measure when the attachment file size exceeds the possible backup size.

File: app/controllers/admin/backups_controller.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    if last_backup
+      @job_status_id = last_backup.job_status.job_id
+      @last_backup_date = I18n.localize(last_backup.updated_at)
+      @last_backup_attachment_id = last_backup.attachments.first&.id
+    end
+  end
+
+  def reset_token
+    if request.post?
```
Comment (oliverguenther) [inline app/controllers/admin/backups_controller.rb]: I generally don't like the `if request.post?` too much. How about two separate routes for get/post with two different actions?

File: frontend/src/app/components/admin/backup.component.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+import { JobStatusModal } from "core-app/modules/job-status/job-status-modal/job-status.modal";
+import { PathHelperService } from 'core-app/modules/common/path-helper/path-helper.service';
+import { OpModalService } from "core-app/modules/modal/modal.service";
+
+export const backupSelector = 'backup';
+
+@Component({
+  selector: backupSelector,
+  templateUrl: './backup.component.html',
+  styleUrls: ['./backup.component.sass']
```
Comment (oliverguenther) [inline frontend/src/app/components/admin/backup.component.ts]: You don't have to create a .sass entry here if you don't have any styles. You can just remove it.

**Review by oliverguenther — CHANGES_REQUESTED** [review]
Only some minor comments code-wise

**Review by oliverguenther — APPROVED** [review]
LGTM :+1: I observed two corner cases where attachments were missing which shouldn't usually happen but I still pushed a miner fix for that.

Also, on my instance the export would exceed the attachment file size, so I added a skip check for the filesize in case the attachment is an export. Since you based Backups on the Export class, this will also fix it for work package or other exports.
